flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/HashBasedCachedBuffer.java

```diff
@@ -0,0 +1,85 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage;
+
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageSubpartitionId;
+import org.apache.flink.util.ExceptionUtils;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+/**
+ * The hash-based mode to accumulate the records. Each subpartition use a {@link
+ * SubpartitionCachedBuffer} to accumulate the received records.
+ *
+ * <p>Note that {@link #setup} need an argument of buffer flush listener to accept the finished
+ * accumulated buffers.
+ */
+public class HashBasedCachedBuffer {
+
+    private final int numSubpartitions;
+
+    private final SubpartitionCachedBuffer[] subpartitionCachedBuffers;
+
+    HashBasedCachedBuffer(
+            int numSubpartitions,
+            int bufferSize,
+            TieredStorageMemoryManager storageMemoryManager,
+            CacheFlushManager cacheFlushManager) {
+        this.numSubpartitions = numSubpartitions;
+        this.subpartitionCachedBuffers = new SubpartitionCachedBuffer[numSubpartitions];
+
+        for (int i = 0; i < numSubpartitions; i++) {
+            subpartitionCachedBuffers[i] =
+                    new SubpartitionCachedBuffer(
+                            new TieredStorageSubpartitionId(i),
+                            bufferSize,
+                            storageMemoryManager,
+                            cacheFlushManager);
+        }
+    }
+
+    public void setup(BiConsumer<TieredStorageSubpartitionId, List<Buffer>> bufferFlusher) {
+        for (int i = 0; i < numSubpartitions; i++) {
+            subpartitionCachedBuffers[i].setup(bufferFlusher);
+        }
+    }
+
+    public void append(
+            ByteBuffer record, TieredStorageSubpartitionId subpartitionId, Buffer.DataType dataType)
+            throws IOException {
+        try {
+            getCachedBuffer(subpartitionId).append(record, dataType);
+        } catch (InterruptedException e) {
+            ExceptionUtils.rethrow(e);
+        }
+    }
+
+    public void close() {
+        Arrays.stream(subpartitionCachedBuffers).forEach(SubpartitionCachedBuffer::close);
+    }
+
+    private SubpartitionCachedBuffer getCachedBuffer(TieredStorageSubpartitionId subpartitionId) {
```

Comment:
I'd suggest the name `getSubpartitionCachedBuffer`.

Reply:
Fixed.